beacon-chain/cache/sync_committee.go

```diff
@@ -0,0 +1,187 @@
+	})
+	// SyncCommitteeCacheHit tracks the number of committee requests that are in the cache.
+	SyncCommitteeCacheHit = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "sync_committee_index_cache_hit_total",
+		Help: "The number of committee requests that are present in the sync committee index cache.",
+	})
+)
+
+// SyncCommitteeCache utilizes a FIFO cache to sufficiently cache validator position within sync committee.
+// It is thread safe with concurrent read write.
+type SyncCommitteeCache struct {
+	cache *cache.FIFO
+	lock  sync.RWMutex
+}
+
+// Index position of all validators in sync committee where `currentSyncCommitteeRoot` is the
+// key and `vIndexToPositionMap` is value. Inside `vIndexToPositionMap`, validator positions
+// are cached where key is the validator index and the value is the `positionInCommittee` struct.
+type syncCommitteeIndexPosition struct {
+	currentSyncCommitteeRoot [32]byte
+	vIndexToPositionMap      map[types.ValidatorIndex]*positionInCommittee
+}
+
+// Index position of individual validator of current period and next period sync committee.
+type positionInCommittee struct {
+	currentPeriod []types.CommitteeIndex
+	nextPeriod    []types.CommitteeIndex
+}
+
+// NewSyncCommittee initializes and returns a new SyncCommitteeCache.
+func NewSyncCommittee() *SyncCommitteeCache {
+	return &SyncCommitteeCache{
+		cache: cache.NewFIFO(keyFn),
+	}
+}
+
+// CurrentPeriodIndexPosition returns current period index position of a validator index with respect with
+// sync committee. If the input validator index has no assignment, an empty list will be returned.
+// If the input root does not exist in cache, `ErrNonExistingSyncCommitteeKey` is returned.
+// Manual checking of state for index position in state is recommended when `ErrNonExistingSyncCommitteeKey` is returned.
+func (s *SyncCommitteeCache) CurrentPeriodIndexPosition(root [32]byte, valIdx types.ValidatorIndex) ([]types.CommitteeIndex, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	pos, err := s.idxPositionInCommittee(root, valIdx)
+	if err != nil {
+		return nil, err
+	}
+	if pos == nil {
+		return []types.CommitteeIndex{}, nil
+	}
+
+	return pos.currentPeriod, nil
+}
+
+// NextPeriodIndexPosition returns next period index position of a validator index in respect with sync committee.
+// If the input validator index has no assignment, an empty list will be returned.
+// If the input root does not exist in cache, `ErrNonExistingSyncCommitteeKey` is returned.
+// Manual checking of state for index position in state is recommended when `ErrNonExistingSyncCommitteeKey` is returned.
+func (s *SyncCommitteeCache) NextPeriodIndexPosition(root [32]byte, valIdx types.ValidatorIndex) ([]types.CommitteeIndex, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	pos, err := s.idxPositionInCommittee(root, valIdx)
+	if err != nil {
+		return nil, err
+	}
+	if pos == nil {
+		return []types.CommitteeIndex{}, nil
+	}
```

Comment:
This block is identical to the contents of CurrentPeriodIndexPosition. Consider refactoring these to share the same code.

Reply:
still pending resolution @terencechain

Reply:
offline discussion there's no easy way to refactor this. "A little duplication is far cheaper than the wrong abstraction"

https://discord.com/channels/476244492043812875/774005284188323870/872154033362726932